File: go-api/src/models/Product.go
package models

type Product struct {
	ID              int32   `json:"id" db:"id"`
	Title           string  `json:"title"`
	BrandId         int32   `json:"brand_id"`
	Description     string  `json:"description"`
	Price           float64 `json:"price"`
	AuthorId        int32   `json:"author_id"`
	SellerID        int32   `json:"seller_id"`
	Discount        int32   `json:"discount"`
	Stock           int32   `json:"stock"`
	Cover           string  `json:"cover"`
	Tags            string  `json:"tags"`
	Attributes      string  `json:"attributes"`
	SpecificationId int32   `json:"specification_id"`
	CreatedAt       string  `json:"created_at"`
	UpdatedAt       string  `json:"updated_at"`
}

type ProductWithAuthor struct {
	Product
	Username string `json:"username"`
	Avatar   string `json:"avatar"`
	Email    string `json:"email"`
}

File: nodejs-api/src/models/Product.ts
import mongoose from "mongoose";
import {ObjectFlags} from "../types";


export interface ProductType{
  _id?: string
  title: string
  brand_id: string
  discount: string
  price: string
  sold: string
  views: string
  author_id: string
  seller_id: string
  stock: number
  cover: string
  tags: string[]
  attributes: object
  details_id: string
  description: string
  createdAt?: Date
  updatedAt?: Date
}


const schemaObject: ObjectFlags<ProductType> = {
  title: {
    type: String,
    unique: true,
    index: true
  },
  cover: String,
  description: String,
  brand_id: {
    index: true,
    type: mongoose.Schema.Types.ObjectId,
    ref: "Brand"
  },
  author_id: {type: mongoose.Schema.Types.ObjectId, ref: "User"},
  seller_id: {type: mongoose.Schema.Types.ObjectId, ref: "User"},
  details_id: {type: mongoose.Schema.Types.ObjectId, ref: "ProductDetails"},
  discount: Number,
  price: Number,
  stock: Number,
  sold: {
    type: Number,
    index: true,
  },
  attributes: {type: Map},
  tags: [String],
  views: Number
}

export default mongoose.model('Product', new mongoose.Schema(schemaObject, {timestamps: true}));
